pkg/planner/core/rule_partition_processor.go

```diff
@@ -181,30 +181,27 @@ func (s *partitionProcessor) getUsedHashPartitions(ctx sessionctx.Context,
 				}
 
 				var rangeScalar uint64
-				var offset int64
 				if mysql.HasUnsignedFlag(col.RetType.GetFlag()) {
 					// Avoid integer overflow
 					if uint64(posHigh) < uint64(posLow) {
 						rangeScalar = 0
 					} else {
 						rangeScalar = uint64(posHigh) - uint64(posLow)
-						offset = int64(uint64(posLow) % uint64(numPartitions))
 					}
 				} else {
 					// Avoid integer overflow
 					if posHigh < posLow {
 						rangeScalar = 0
 					} else {
 						rangeScalar = uint64(posHigh - posLow)
-						offset = posLow % int64(numPartitions)
 					}
 				}
 
 				// if range is less than the number of partitions, there will be unused partitions we can prune out.
 				if rangeScalar < uint64(numPartitions) && !highIsNull && !lowIsNull {
 					var i int64
 					for i = 0; i <= int64(rangeScalar); i++ {
-						idx := mathutil.Abs(offset+i) % int64(numPartitions)
+						idx := mathutil.Abs((posLow + i) % int64(numPartitions))
```

Comment:
Ref https://github.com/pingcap/tidb/pull/49853/files#diff-4990f33f249ec8823358eef763ccfccd9284b58d4942eba35b7b4e22c16b1a33L193, let it same as before.

Although I think the behaviour of hash function is strange, it already works for a long time. Just follow it.